### PR TITLE
[NEX-4] Authentication

### LINKED ISSUE #4: [NEX-4] Authentication

### PR DESCRIPTION
**Pull request by stodorichkov**

### CODE REVIEW

File: nexus/src/main/java/com/example/nexus/config/JwtAuthEntryPoint.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.example.nexus.config;
+
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.stereotype.Component;
+
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/JwtAuthEntryPoint.java]: Remove blank line

File: nexus/src/main/java/com/example/nexus/config/JwtFilter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/JwtFilter.java]: Remove blank line.

File: nexus/src/main/java/com/example/nexus/config/SecurityConfig.java [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 import com.example.nexus.constant.RequestMappingConstants;
 import com.example.nexus.constant.RoleConstants;
+
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/SecurityConfig.java]: Remove blank line.

File: nexus/src/main/java/com/example/nexus/config/SecurityConfig.java [hunk 4]
```diff
@@ -10,24 +12,36 @@
...
                         .anyRequest().authenticated()
                 )
-                .httpBasic(Customizer.withDefaults())
-                .build();
+                .httpBasic(Customizer.withDefaults());
+
+        http.addFilterBefore(jwtFilter, UsernamePasswordAuthenticationFilter.class);
+
+        return http.build();
+
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/SecurityConfig.java]: Remove blank line.

File: nexus/src/main/java/com/example/nexus/config/SecurityConfig.java [hunk 5]
```diff
@@ -10,24 +12,36 @@
...
     public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
-        return http
+        http
                 .csrf(AbstractHttpConfigurer::disable)
+                .exceptionHandling(ex -> ex.authenticationEntryPoint(authEntryPoint))
+                .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                 .authorizeHttpRequests(auth -> auth
                         .requestMatchers(RequestMappingConstants.AUTH).permitAll()
-                        .requestMatchers(RequestMappingConstants.ADMIN).hasRole(RoleConstants.ADMIN)
+                        .requestMatchers(RequestMappingConstants.ADMIN).hasAuthority(RoleConstants.ADMIN)
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/SecurityConfig.java:35]: Remove blank line

File: nexus/src/main/java/com/example/nexus/config/SecurityConfig.java [hunk 6]
```diff
@@ -10,24 +12,36 @@
...
-                .build();
+                .httpBasic(Customizer.withDefaults());
+
+        http.addFilterBefore(jwtFilter, UsernamePasswordAuthenticationFilter.class);
+
+        return http.build();
+
     }
 
     @Bean
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/config/SecurityConfig.java:45]: Looks like this annotation points to nothing and should be removed.

File: nexus/src/main/java/com/example/nexus/service/JwtServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.security.Keys;
+import jakarta.servlet.http.HttpServletRequest;
+import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/service/JwtServiceImpl.java]: Remove blank line

File: nexus/src/main/java/com/example/nexus/service/RoleServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class RoleServiceImpl implements RoleService {
+    private final RoleRepository roleRepository;
+
+    @Override
+    public void seedRole(String name) {
+        if(this.roleRepository.findByName(name).isPresent()) {
+            return;
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/service/RoleServiceImpl.java:16]: Maybe add some kind of logging above return for debugging purposes.

File: nexus/src/main/java/com/example/nexus/service/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+public class UserServiceImpl implements UserService {
+    private final ProfileRepository profileRepository;
+    private final UserRepository userRepository;
+    private final RoleRepository roleRepository;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    public void seedAdmin() {
+        if(userRepository.findByUsername(AdminConstants.USERNAME).isPresent()) {
+            return;
```
Comment (LuckyOx04) [inline nexus/src/main/java/com/example/nexus/service/UserServiceImpl.java:24]: Here there may also be good to add some logging.

File: nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Optional;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java]: Remove blank line.

File: nexus/src/test/java/com/example/nexus/service/JwtServiceImplTests.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
+import org.springframework.security.authentication.TestingAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/JwtServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/JwtServiceImplTests.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.List;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/JwtServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/RoleServiceImplTests.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.example.nexus.model.entity.Role;
+import com.example.nexus.repository.RoleRepository;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/RoleServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/RoleServiceImplTests.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Optional;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/RoleServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/UserDetailsServiceImplTests.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+import com.example.nexus.model.entity.User;
+import com.example.nexus.repository.UserRepository;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/UserDetailsServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/UserDetailsServiceImplTests.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Optional;
+import java.util.Set;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/UserDetailsServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/UserServiceImplTests.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+import com.example.nexus.repository.UserRepository;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/UserServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/UserServiceImplTests.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/UserServiceImplTests.java]: Remove blank line

File: nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    @Test
+    void login_invalidPassword_expectUnauthorizedException() {
+        when(this.userRepository.findByUsername(request.username())).thenReturn(Optional.of(new User()));
+
+        assertThrows(UnauthorizedException.class, () -> this.authService.login(request));
+    }
+
+    @Test
+    void login_userExist_expectToken() {
+        when(this.userRepository.findByUsername(request.username())).thenReturn(Optional.of(user));
```
Comment (LuckyOx04) [inline nexus/src/test/java/com/example/nexus/service/AuthServiceImplTests.java:75]: Remove blank line

File: nexus/src/main/java/com/example/nexus/config/SecurityConfig.java (hunk 5 above)
Comment (stodorichkov) [reply nexus/src/main/java/com/example/nexus/config/SecurityConfig.java:35]: It is separate filter and must be on new line

**Review by LuckyOx04 — APPROVED** [review]
Everything is fine.

**Pull request merged**